README.md
简体中文 | [English](./README_en.md)

<p align="center">
  <img src="./docs/imgs/paddlenlp.png" width="718" height ="100" />
</p>

------------------------------------------------------------------------------------------
[![PyPI - PaddleNLP Version](https://img.shields.io/pypi/v/paddlenlp.svg?label=pip&logo=PyPI&logoColor=white)](https://pypi.org/project/paddlenlp/)
[![PyPI - Python Version](https://img.shields.io/pypi/pyversions/paddlenlp)](https://pypi.org/project/paddlenlp/)
[![PyPI Status](https://pepy.tech/badge/paddlenlp/month)](https://pepy.tech/project/paddlenlp)
![support os](https://img.shields.io/badge/os-linux%2C%20win%2C%20mac-yellow.svg)
![GitHub](https://img.shields.io/github/license/paddlepaddle/paddlenlp)

## News  <img src="./docs/imgs/news_icon.png" width="40"/>

* [2021-12-28] PaddleNLP新发**语义检索、问答、评论观点抽取和情感倾向分析** [产业化案例](https://github.com/PaddlePaddle/PaddleNLP/tree/develop/applications)，🚀快速搭建系统！配套视频课程[直通车](https://aistudio.baidu.com/aistudio/course/introduce/24902)！
* [2021-12-12] PaddleNLP v2.2版本已发布！:tada: 欢迎体验更快的文本处理[FasterTokenizer](./examples/faster/faster_ernie)、更快的预训练模型[FasterERNIE](./examples/faster/faster_ernie)、更快的文本生成[FasterGeneration](./examples/faster/faster_generation)；新推出『解语』名词短语标注工具[NPTag](./examples/text_to_knowledge/nptag)、超快中文小模型[PP-MiniLM](./examples/model_compression/pp-minilm)！ 更多详细升级信息请查看[Release Note](https://github.com/PaddlePaddle/PaddleNLP/releases/tag/v2.1.0)。
* [2021-12-12] 飞桨新产品**端到端问答工具**🚀[RocketQA](https://github.com/PaddlePaddle/RocketQA)全新发布！:tada:

## 简介

PaddleNLP是飞桨自然语言处理开发库，具备**易用的文本领域API**，**多场景的应用示例**、和**高性能分布式训练**三大特点，旨在提升开发者在文本领域的开发效率，并提供丰富的NLP应用示例。

- **易用的文本领域API**
  - 提供丰富的产业级预置任务能力[Taskflow](./docs/model_zoo/taskflow.md)和全流程的文本领域API：支持丰富中文数据集加载的[Dataset API](https://paddlenlp.readthedocs.io/zh/latest/data_prepare/dataset_list.html)；灵活高效地完成数据预处理的[Data API](https://paddlenlp.readthedocs.io/zh/latest/source/paddlenlp.data.html)；提供100+预训练模型的[Transformers API](./docs/model_zoo/transformers.rst)等，可大幅提升NLP任务建模的效率。

- **多场景的应用示例**
  - 覆盖从学术到产业级的NLP[应用示例](#多场景的应用示例)，涵盖NLP基础技术、NLP系统应用以及相关拓展应用。全面基于飞桨核心框架2.0全新API体系开发，为开发者提供飞桨文本领域的最佳实践。

- **高性能分布式训练**
  - 基于飞桨核心框架领先的自动混合精度优化策略，结合分布式Fleet API，支持4D混合并行策略，可高效地完成大规模预训练模型训练。

## 安装

### 环境依赖

- python >= 3.6
- paddlepaddle >= 2.2

### pip安装

```shell
pip install --upgrade paddlenlp
```

更多关于PaddlePaddle和PaddleNLP安装的详细教程请查看[Installation](./docs/get_started/installation.rst)。

## 易用的文本领域API

### Taskflow：开箱即用的产业级NLP能力

Taskflow旨在提供**开箱即用**的NLP预置任务能力，覆盖自然语言理解与生成两大场景，提供**产业级的效果**与**极致的预测性能**。

```python
from paddlenlp import Taskflow

# 中文分词
seg = Taskflow("word_segmentation")
seg("第十四届全运会在西安举办")
>>> ['第十四届', '全运会', '在', '西安', '举办']

# 词性标注
tag = Taskflow("pos_tagging")
tag("第十四届全运会在西安举办")
>>> [('第十四届', 'm'), ('全运会', 'nz'), ('在', 'p'), ('西安', 'LOC'), ('举办', 'v')]

# 命名实体识别
ner = Taskflow("ner")
ner("《孤女》是2010年九州出版社出版的小说，作者是余兼羽")
>>> [('《', 'w'), ('孤女', '作品类_实体'), ('》', 'w'), ('是', '肯定词'), ('2010年', '时间类'), ('九州出版社', '组织机构类'), ('出版', '场景事件'), ('的', '助词'), ('小说', '作品类_概念'), ('，', 'w'), ('作者', '人物类_概念'), ('是', '肯定词'), ('余兼羽', '人物类_实体')]

# 『解语』- 名词短语标注
nptag = Taskflow("knowledge_mining", model="nptag")
nptag("红曲霉菌")
>>> [{'text': '红曲霉菌', 'label': '微生物'}]

# 句法分析
ddp = Taskflow("dependency_parsing")
ddp("9月9日上午纳达尔在亚瑟·阿什球场击败俄罗斯球员梅德韦杰夫")
>>> [{'word': ['9月9日', '上午', '纳达尔', '在', '亚瑟·阿什球场', '击败', '俄罗斯', '球员', '梅德韦杰夫'], 'head': [2, 6, 6, 5, 6, 0, 8, 9, 6], 'deprel': ['ATT', 'ADV', 'SBV', 'MT', 'ADV', 'HED', 'ATT', 'ATT', 'VOB']}]

# 情感分析
senta = Taskflow("sentiment_analysis")
senta("这个产品用起来真的很流畅，我非常喜欢")
>>> [{'text': '这个产品用起来真的很流畅，我非常喜欢', 'label': 'positive', 'score': 0.9938690066337585}]
```
更多使用方法请参考[Taskflow文档](./docs/model_zoo/taskflow.md)。

### Transformers API: 强大的预训练模型生态底座

覆盖**30**个网络结构和**100**余个预训练模型参数，既包括百度自研的预训练模型如ERNIE系列, PLATO, SKEP等，也涵盖业界主流的中文预训练模型如BERT，GPT，XLNet，BART等。使用AutoModel可以下载不同网络结构的预训练模型。欢迎开发者加入贡献更多预训练模型！🤗

```python
from paddlenlp.transformers import *

ernie = AutoModel.from_pretrained('ernie-1.0')
ernie_gram = AutoModel.from_pretrained('ernie-gram-zh')
bert = AutoModel.from_pretrained('bert-wwm-chinese')
albert = AutoModel.from_pretrained('albert-chinese-tiny')
roberta = AutoModel.from_pretrained('roberta-wwm-ext')
electra = AutoModel.from_pretrained('chinese-electra-small')
gpt = AutoModelForPretraining.from_pretrained('gpt-cpm-large-cn')
```

对预训练模型应用范式如语义表示、文本分类、句对匹配、序列标注、问答等，提供统一的API体验。

```python
import paddle
from paddlenlp.transformers import *

tokenizer = AutoTokenizer.from_pretrained('ernie-1.0')
text = tokenizer('自然语言处理')

# 语义表示
model = AutoModel.from_pretrained('ernie-1.0')
sequence_output, pooled_output = model(input_ids=paddle.to_tensor([text['input_ids']]))
# 文本分类 & 句对匹配
model = AutoModelForSequenceClassification.from_pretrained('ernie-1.0')
# 序列标注
model = AutoModelForTokenClassification.from_pretrained('ernie-1.0')
# 问答
model = AutoModelForQuestionAnswering.from_pretrained('ernie-1.0')
```

请参考[Transformer API文档](./docs/model_zoo/transformers.rst)查看目前支持的预训练模型结构、参数和详细用法。

### Datasets API: 丰富的中文数据集

Dataset API提供便捷、高效的数据集加载功能；内置[千言数据集](https://www.luge.ai/)，提供丰富的面向自然语言理解与生成场景的中文数据集，为NLP研究人员提供一站式的科研体验。

```python
from paddlenlp.datasets import load_dataset

train_ds, dev_ds, test_ds = load_dataset("chnsenticorp", splits=["train", "dev", "test"])

train_ds, dev_ds = load_dataset("lcqmc", splits=["train", "dev"])
```

可参考[Dataset文档](https://paddlenlp.readthedocs.io/zh/latest/data_prepare/dataset_list.html) 查看更多数据集。

### Embedding API: 一键加载预训练词向量

```python
from paddlenlp.embeddings import TokenEmbedding

wordemb = TokenEmbedding("w2v.baidu_encyclopedia.target.word-word.dim300")
print(wordemb.cosine_sim("国王", "王后"))
>>> 0.63395125
wordemb.cosine_sim("艺术", "火车")
>>> 0.14792643
```

内置50+中文词向量，覆盖多种领域语料、如百科、新闻、微博等。更多使用方法请参考[Embedding文档](./docs/model_zoo/embeddings.md)。

### 更多API使用文档

- [Data API](./docs/data.md): 提供便捷高效的文本数据处理功能
- [Metrics API](./docs/metrics.md): 提供NLP任务的评估指标，与飞桨高层API兼容。

更多的API示例与使用说明请查阅[PaddleNLP官方文档](https://paddlenlp.readthedocs.io/)

## 多场景的应用示例

PaddleNLP提供了多粒度、多场景的NLP应用示例，面向动态图模式和全新的API体系开发，更加简单易懂。
涵盖了[NLP基础技术](#nlp-基础技术)、[NLP系统应用](#nlp-系统应用)以及文本相关的[NLP拓展应用](#拓展应用)、与知识库结合的[文本知识关联](./examples/text_to_knowledge)、与图结合的[文本图学习](./examples/text_graph/)等。

### NLP 基础技术

| 任务   | 简介     |
| :------------  | ---- |
| [词向量](./examples/word_embedding/) | 利用`TokenEmbedding API`展示如何快速计算词之间语义距离和词的特征提取。 |
| [词法分析](./examples/lexical_analysis/) | 基于BiGRU-CRF模型实现了分词、词性标注和命名实体识的联合训练任务。 |
| [语言模型](./examples/language_model/)  | 覆盖了经典的LSTM类语言模型，和Transformer类的预训练语言模型，如BERT, ERNIE, GPT, XLNet, BART，ELECTRA，BigBird等。 |
| [语义解析](./examples/text_to_sql/):star: | 语义解析Text-to-SQL任务是让机器自动让自然语言问题转换数据库可操作的SQL查询语句，是实现基于数据库自动问答的核心模块。|

#### 句法依存分析 (Dependency Parsing)

| 模型  | 简介    |
| :----- | ------ |
| [Deep Biaffine Parsing](./examples/dependency_parsing/) | 实现经典的Deep Biaffine Parsing句法分析模型，并提供高质量的中文预训练模型可供直接使用。|

#### 文本分类 (Text Classification)

| 模型  | 简介    |
| :----- | ------ |
| [RNN/CNN/GRU/LSTM](./examples/text_classification/rnn) | 实现了经典的RNN, CNN, GRU, LSTM等经典文本分类结构。|
| [BiLSTM-Attention](./examples/text_classification/rnn) | 基于BiLSTM网络结构引入注意力机制提升文本分类效果。 |
| [BERT/ERNIE](./examples/text_classification/pretrained_models) | 提供基于预训练模型的文本分类任务实现，包含训练、预测和推理部署的全流程应用。 |

#### 文本匹配 (Text Matching)

| 模型    | 简介       |
| :--------------- | ---------- |
| [SimCSE](./examples/text_matching/simcse/):star2:  | 基于论文[SimCSE: Simple Contrastive Learning of Sentence Embeddings](https://arxiv.org/abs/2104.08821)实现无监督语义匹配模型，无需标注数据仅利用无监督数据也能训练效果出众的语义匹配模型。|
| [ERNIE-Gram w/ R-Drop](./examples/text_matching/question_matching/) | 提供基于ERNIE-Gram预训练模型结合R-Drop策略的问题匹配任在千言数据集上的基线代码。|
| [SimNet](./examples/text_matching/simnet/)  | 百度自研的语义匹配框架，使用BOW、CNN、GRNN等核心网络作为表示层，在百度内搜索、推荐等多个应用场景得到广泛易用。|
| [ERNIE](./examples/text_matching/ernie_matching/) | 基于ERNIE使用LCQMC数据完成中文句对匹配任务，提供了Pointwise和Pairwise两种类型学习方式。 |
| [Sentence-BERT](./examples/text_matching/sentence_transformers/) | 提供基于Siamese双塔结构的文本匹配模型[Sentence-BERT](https://arxiv.org/abs/1908.1008)实现，可用于获取文本的向量化表示。 |
| [SimBERT](./examples/text_matching/simbert/) | 提供[SimBERT](https://github.com/ZhuiyiTechnology/simbert)模型实现，用于获取文本的向量化表示。|

#### 文本生成 (Text Generation)

| 模型        | 简介      |
| :------------ | ---------- |
| [Seq2Seq](./examples/text_generation/couplet) | 实现了经典的Seq2Seq with Attention的网络结构，并提供在自动对联的文本生成应用示例。 |
| [VAE-Seq2Seq](./examples/text_generation/vae-seq2seq) | 在Seq2Seq框架基础上，加入VAE结构以实现更加多样化的文本生成。|
| [ERNIE-GEN](./examples/text_generation/ernie-gen) | [ERNIE-GEN](https://arxiv.org/abs/2001.11314)是百度NLP提出的基于多流(multi-flow)机制生成完整语义片段的预训练模型，基于该模型实现了提供了智能写诗的应用示例。|

#### 文本纠错 (Text Correction)

| 模型        | 简介      |
| :------------ | ---------- |
| [ERNIE-CSC](./examples/text_correction/ernie-csc) :star2:  | [ERNIE-CSC](https://aclanthology.org/2021.findings-acl.198.pdf)是基于ERNIE预训练模型融合了拼音特征的端到端中文拼写纠错模型，在SIGHAN数据集上取得SOTA的效果。|

#### 语义索引 (Semantic Indexing)

提供一套完整的语义索引开发流程，并提供了In-Batch Negative和Hardest Negatives两种策略，开发者可基于该示例实现一个轻量级的语义索引系统，更多信息请查看[语义索引应用示例](./examples/semantic_indexing/)。

#### 信息抽取 (Information Extraction)

| 任务   | 简介     |
| :---------------  | ---- |
| [DuEE](./examples/information_extraction/DuEE/) | 基于[DuEE](https://link.springer.com/chapter/10.1007/978-3-030-60457-8_44)数据集，使用预训练模型的方式提供句子级和篇章级的事件抽取示例。 |
| [DuIE](./examples/information_extraction/DuIE/) | 基于[DuIE](http://tcci.ccf.org.cn/conference/2019/papers/EV10.pdf)数据集，使用预训练模型的方式提供关系抽取示例。 |
| [快递单信息抽取](./examples/information_extraction/waybill_ie/) | 提供BiLSTM+CRF和预训练模型两种方式完成真实的快递单信息抽取案例。 |

### NLP 系统应用

#### 情感分析 (Sentiment Analysis)

| 模型      | 简介       |
| :--------- | ---------- |

docs/changelog.md
### v2.0.2 (2021.06.07)

- 丰富预训练模型
  * 新增多粒度语言知识预训练模型[ERNIE-Gram](https://arxiv.org/abs/2010.12148)，该模型在多项中文NLP任务取得SOTA成绩。
  * 新增NeZha中文预训练模型，感谢 @jm12138 的高质量贡献！ 🎉 🎉 🎉
  * 新增GPT CPM-Distill中文小型化模型，感谢 @jm12138 的高质量贡献！🎉 🎉 🎉

- Bug Fix
  * 修复了`softmax_with_crossentropy` API导致的deprecated warning
  * 更正了`ChnSentiCorp`, `LCQMC`等数据集的官方下载链接。

### v2.0.1 (2021.05.21)

- Bug Fix
  - 修复Windows CPU环境下的import产生的CUDA_HOME检测问题。

### v2.0.0 (2021.05.20)

PaddleNLP 2.0是飞桨生态的文本领域核心库，具备**易用的文本领域API**，**多场景的应用示例**、和**高性能分布式训练**三大特点，旨在提升飞桨开发者文本领域建模效率，并提供基于飞桨框架2.0的NLP领域最佳实践。

### 特性

- **易用的文本领域API**
  - 提供从数据集加载、文本预处理、组网建模、评估、到推的领域API：如一键加载丰富中文数据集的Dataset API, 可灵活高效的进行数据与处理的Data API，预置60+预训练词向量的Embedding API, 内置50+预训练模型，提供预训练模型生态基础设施的Transformer API等，可大幅提升NLP任务建模和迭代的效率。更多API详细说明请查看[PaddleNLP官方文档](https://paddlenlp.readthedocs.io/)


- **多场景的应用示例**
  - PaddleNLP 2.0提供多粒度多场景的应用示例，涵盖从NLP基础技术、NLP核心技术、NLP系统应用以及文本相关的拓展应用等。全面基于飞桨2.0全新API体系开发，为开发提供飞桨2.0框架在文本领域的最佳实践。


- **高性能分布式训练**
  - 基于飞桨核心框架『**动静统一**』的特性与领先的自动混合精度优化策略，通过分布式Fleet API，支持超大规模参数的4D混合并行策略，并且可根据硬件情况灵活可配，高效地完成超大规模参数的模型训练。


compare.py
# coding:utf-8
import sys
from paddlenlp.transformers import (
    RobertaModel, RobertaForMaskedLM, RobertaForQuestionAnswering,
    RobertaForSequenceClassification, RobertaForTokenClassification)
from paddlenlp.transformers import RobertaBPETokenizer, RobertaTokenizer
import paddle
import os
import numpy as np


def decode(start, end, topk, max_answer_len, undesired_tokens):
    """
        Take the output of any :obj:`ModelForQuestionAnswering` and will generate probabilities for each span to be the
        actual answer.
        """
    # Ensure we have batch axis
    if start.ndim == 1:
        start = start[None]

    if end.ndim == 1:
        end = end[None]
    # Compute the score of each tuple(start, end) to be the real answer
    outer = np.matmul(np.expand_dims(start, -1), np.expand_dims(end, 1))

    # Remove candidate with end < start and end - start > max_answer_len
    candidates = np.tril(np.triu(outer), max_answer_len - 1)

    #  Inspired by Chen & al. (https://github.com/facebookresearch/DrQA)
    scores_flat = candidates.flatten()
    if topk == 1:
        idx_sort = [np.argmax(scores_flat)]
    elif len(scores_flat) < topk:
        idx_sort = np.argsort(-scores_flat)
    else:
        idx = np.argpartition(-scores_flat, topk)[0:topk]
        idx_sort = idx[np.argsort(-scores_flat[idx])]

    starts, ends = np.unravel_index(idx_sort, candidates.shape)[1:]
    desired_spans = np.isin(starts, undesired_tokens.nonzero()) & np.isin(
        ends, undesired_tokens.nonzero())
    starts = starts[desired_spans]
    ends = ends[desired_spans]
    scores = candidates[0, starts, ends]

    return starts, ends, scores


def compare_roberta_base(model_name='roberta-base'):
    path = 'paddlenlp/transformers/roberta/{}'.format(model_name.split('/')[-1])

    model = RobertaForMaskedLM.from_pretrained(path)
    tokenizer = RobertaBPETokenizer.from_pretrained(model_name)
    text = ["The man worked as a", "."]  #"The man worked as a <mask>."
    tokens_list = []
    for i in range(2):
        tokens_list.append(tokenizer.tokenize(text[i]))

    tokens = ['<s>']
    tokens.extend(tokens_list[0])
    tokens.extend(['<mask>'])
    tokens.extend(tokens_list[1])
    tokens.extend(['</s>'])
    token_ids = tokenizer.convert_tokens_to_ids(tokens)
    # print(token_ids)

    model.eval()
    input_ids = paddle.to_tensor([token_ids])
    with paddle.no_grad():
        pd_outputs = model(input_ids)

    pd_outputs = pd_outputs[0]

    pd_outputs_sentence = "paddle: "
    for i, id in enumerate(token_ids):
        if id == 50264:
            scores, index = paddle.nn.functional.softmax(pd_outputs[i],
                                                         -1).topk(5)
            tokens = tokenizer.convert_ids_to_tokens(index.tolist())
            outputs = []
            for score, tk in zip(scores.tolist(), tokens):
                outputs.append(f"{tk}={score}")
            pd_outputs_sentence += "[" + "||".join(outputs) + "]" + " "
        else:
            pd_outputs_sentence += "".join(
                tokenizer.convert_ids_to_tokens(
                    [id], skip_special_tokens=True)) + " "

    print(pd_outputs_sentence)


def compare_deepsort_qa():
    path = 'paddlenlp/transformers/roberta/roberta-base-squad2'
    tokenizer = RobertaBPETokenizer.from_pretrained("roberta-base")
    questions = ['Where do I live?']
    contexts = ['My name is Sarah and I live in London']

    token = tokenizer(
        questions,
        contexts,
        stride=128,
        max_seq_len=64,
        return_attention_mask=True,
        return_special_tokens_mask=True)
    # print(token)
    special_tokens_mask = token[0]['special_tokens_mask']
    count = 3
    st_idx = 0
    for i in special_tokens_mask:
        st_idx += 1
        if i == 1:
            count -= 1
        if count == 0:
            break

    input_ids = token[0]['input_ids']
    offset_mapping = token[0]['offset_mapping']

    input_ids = paddle.to_tensor(input_ids, dtype='int64').unsqueeze(0)

    model = RobertaForQuestionAnswering.from_pretrained(path)
    model.eval()
    start, end = model(input_ids=input_ids)
    start_ = start[0].numpy()
    end_ = end[0].numpy()
    undesired_tokens = np.ones_like(input_ids[0].numpy())

    undesired_tokens[1:st_idx] = 0
    undesired_tokens[-1] = 0

    # Generate mask
    undesired_tokens_mask = undesired_tokens == 0.0

    # Make sure non-context indexes in the tensor cannot contribute to the softmax
    start_ = np.where(undesired_tokens_mask, -10000.0, start_)
    end_ = np.where(undesired_tokens_mask, -10000.0, end_)

    start_ = np.exp(start_ - np.log(
        np.sum(np.exp(start_), axis=-1, keepdims=True)))
    end_ = np.exp(end_ - np.log(np.sum(np.exp(end_), axis=-1, keepdims=True)))
    start_idx, end_idx, score = decode(start_, end_, 1, 64, undesired_tokens)
    start_idx, end_idx = offset_mapping[start_idx[0]][0], offset_mapping[
        end_idx[0]][1]
    print("ans: {}".format(contexts[0][start_idx:end_idx]),
          'score:{}'.format(score.item()))


def compare_uer_text_cls():
    tokenizer = RobertaTokenizer.from_pretrained(
        "uer/roberta-base-finetuned-chinanews-chinese")
    token = tokenizer('北京上个月召开了两会')
    # print(token)
    path = 'paddlenlp/transformers/roberta/roberta-base-finetuned-chinanews-chinese'
    config = RobertaModel.pretrained_init_configuration[
        'uer/roberta-base-finetuned-chinanews-chinese']
    roberta = RobertaModel(**config)
    model = RobertaForSequenceClassification(roberta, 7)
    model_state = paddle.load(os.path.join(path, "model_state.pdparams"))
    model.load_dict(model_state)
    model.eval()
    input_ids = paddle.to_tensor(token['input_ids'], dtype='int64').unsqueeze(0)
    with paddle.no_grad():
        output = model(input_ids)
    import paddle.nn.functional as F
    output = F.softmax(output)
    id2label = {
        "0": "mainland China politics",
        "1": "Hong Kong - Macau politics",
        "2": "International news",
        "3": "financial news",
        "4": "culture",
        "5": "entertainment",
        "6": "sports"
    }
    for i in range(7):
        print("{}: \t {}".format(id2label[str(i)], output[0][i].item()))


def compare_uer_clue_ner():
    text = '江苏警方通报特斯拉冲进店铺'
    tokenizer = RobertaTokenizer.from_pretrained(
        "uer/roberta-base-finetuned-cluener2020-chinese")
    token = tokenizer(text)

    path = 'paddlenlp/transformers/roberta/roberta-base-finetuned-cluener2020-chinese'
    config = RobertaModel.pretrained_init_configuration[
        'uer/roberta-base-finetuned-cluener2020-chinese']
    roberta = RobertaModel(**config)
    model = RobertaForTokenClassification(roberta, 32)
    model_state = paddle.load(os.path.join(path, "model_state.pdparams"))
    model.load_dict(model_state)
    model.eval()
    input_ids = paddle.to_tensor(token['input_ids'], dtype='int64').unsqueeze(0)
    with paddle.no_grad():
        output = model(input_ids)

    import paddle.nn.functional as F
    output = F.softmax(output)
    id2label = {
        "0": "O",
        "1": "B-address",
        "2": "I-address",
        "3": "B-book",
        "4": "I-book",
        "5": "B-company",
        "6": "I-company",
        "7": "B-game",
        "8": "I-game",
        "9": "B-government",
        "10": "I-government",
        "11": "B-movie",
        "12": "I-movie",
        "13": "B-name",
        "14": "I-name",
        "15": "B-organization",
        "16": "I-organization",
        "17": "B-position",
        "18": "I-position",
        "19": "B-scene",
        "20": "I-scene",
        "21": "S-address",
        "22": "S-book",
        "23": "S-company",
        "24": "S-game",
        "25": "S-government",
        "26": "S-movie",
        "27": "S-name",
        "28": "S-organization",
        "29": "S-position",
        "30": "S-scene",
        "31": "[PAD]"
    }
    tokenized_text = tokenizer.tokenize(text)
    scores = []
    char_cn = []
    for t, s in zip(tokenized_text, output[0][1:-1]):
        index = paddle.argmax(s).item()
        label = id2label[str(index)]
        if index != 0:
            scores.append(s[index].item())
            char_cn.append(t)
            print(f"{label} {t} score {s[index].item()}")
    print("{}:{}".format("".join(char_cn[:2]), sum(scores[:2]) / 2))
    print("{}:{}".format("".join(char_cn[2:]), sum(scores[2:]) / 3))


def compare_uer_qa():
    tokenizer = RobertaTokenizer.from_pretrained(
        "uer/roberta-base-chinese-extractive-qa")
    questions = ['著名诗歌《假如生活欺骗了你》的作者是']
    contexts = [
        '普希金从那里学习人民的语言，吸取了许多有益的养料，这一切对普希金后来的创作产生了很大的影响。这两年里，普希金创作了不少优秀的作品，如《囚徒》、《致大海》、《致凯恩》和《假如生活欺骗了你》等几十首抒情诗，叙事诗《努林伯爵》，历史剧《鲍里斯·戈都诺夫》，以及《叶甫盖尼·奥涅金》前六章。'
    ]
    token = tokenizer(questions, contexts, stride=256, max_seq_len=256)
    token_type_ids = token[0]["token_type_ids"]
    offset_mapping = token[0]["offset_mapping"]
    st_idx = len(token_type_ids) - sum(token_type_ids)

    path = 'paddlenlp/transformers/roberta/roberta-base-chinese-extractive-qa'
    model = RobertaForQuestionAnswering.from_pretrained(path)
    model.eval()
    input_ids = paddle.to_tensor(
        token[0]['input_ids'], dtype='int64').unsqueeze(0)
    token_type_ids = paddle.to_tensor(
        token[0]['token_type_ids'], dtype='int64').unsqueeze(0)
    with paddle.no_grad():
        start, end = model(input_ids, token_type_ids)
    start_ = start[0].numpy()
    end_ = end[0].numpy()
    undesired_tokens = np.ones_like(input_ids[0].numpy())

    undesired_tokens[1:st_idx] = 0
    undesired_tokens[-1] = 0

    # Generate mask
    undesired_tokens_mask = undesired_tokens == 0.0

    # Make sure non-context indexes in the tensor cannot contribute to the softmax
    start_ = np.where(undesired_tokens_mask, -10000.0, start_)
    end_ = np.where(undesired_tokens_mask, -10000.0, end_)

    start_ = np.exp(start_ - np.log(
        np.sum(np.exp(start_), axis=-1, keepdims=True)))
    end_ = np.exp(end_ - np.log(np.sum(np.exp(end_), axis=-1, keepdims=True)))
    start_idx, end_idx, score = decode(start_, end_, 1, 64, undesired_tokens)
    start_idx, end_idx = offset_mapping[start_idx[0]][0], offset_mapping[
        end_idx[0]][1]
    print("ans: {}".format(contexts[0][start_idx:end_idx]),
          'score:{}'.format(score.item()))


if __name__ == "__main__":
    # compare_deepsort_qa()
    """
    ### paddlenlp
    ans: London score:0.7771294116973877
    ###huggingface
    https://huggingface.co/deepset/roberta-base-squad2?context=My+name+is+Sarah+and+I+live+in+London&question=Where+do+I+live%3F
    {
    "score": 0.777230978012085,
    "start": 31,
    "end": 37,
    "answer": "London"
    }
    """
    compare_uer_text_cls()
    """
    ### paddlenlp:
    mainland China politics:         0.7211663722991943
    Hong Kong - Macau politics:      0.0015174386790022254
    International news:      0.0003004991449415684
    financial news:          0.19014376401901245
    culture:         0.07077095657587051
    entertainment:   0.007615785114467144
    sports:          0.008485240861773491

    ### huggingface:
    https://huggingface.co/uer/roberta-base-finetuned-chinanews-chinese?text=%E5%8C%97%E4%BA%AC%E4%B8%8A%E4%B8%AA%E6%9C%88%E5%8F%AC%E5%BC%80%E4%BA%86%E4%B8%A4%E4%BC%9A
    [
        {
        "label": "mainland China politics",
        "score": 0.7211663126945496
        },
        {
        "label": "Hong Kong - Macau politics",
        "score": 0.0015174405416473746
        },
        {
        "label": "International news",
        "score": 0.0003004991449415684
        },
        {
        "label": "financial news",
        "score": 0.19014358520507812
        },
        {
        "label": "culture",
        "score": 0.07077110558748245
        },
        {
        "label": "entertainment",
        "score": 0.007615773007273674
        },
        {
        "label": "sports",
        "score": 0.008485235273838043
        }
    ]
    """

    compare_uer_clue_ner()
    """
    ### paddlenlp:
    B-address 江 score 0.6619006395339966
    I-address 苏 score 0.5544567704200745
    B-company 特 score 0.42272716760635376
    I-company 斯 score 0.45469826459884644
    I-company 拉 score 0.5207838416099548
    江苏:0.6081787049770355
    特斯拉:0.466069757938385
    ### hugging_face:
    https://huggingface.co/uer/roberta-base-finetuned-cluener2020-chinese?text=%E6%B1%9F%E8%8B%8F%E8%AD%A6%E6%96%B9%E9%80%9A%E6%8A%A5%E7%89%B9%E6%96%AF%E6%8B%89%E5%86%B2%E8%BF%9B%E5%BA%97%E9%93%BA
    {
        "entity_group": "address",
        "score": 0.6081769466400146,
        "word": "江 苏",
        "start": 0,
        "end": 2
    },
    {
        "entity_group": "company",
        "score": 0.466068834066391,
        "word": "特 斯 拉",
        "start": 6,
        "end": 9
    }
    """
    compare_uer_qa()
    """
    paddlenlp:
    ans: 普希金 score:0.9766426682472229
    huggingface:
    {
    "score": 0.9766426086425781,
    "start": 0,
    "end": 3,
    "answer": "普希金"
    }
    """
    compare_roberta_base('roberta-base')
    """
    ### paddlenlp:
    paddle:  The Ġman Ġworked Ġas Ġa [Ġmechanic=0.08702456951141357||Ġwaiter=0.08196529746055603||Ġbutcher=0.07332348823547363||Ġminer=0.04632236436009407||Ġguard=0.04015030339360237] .

    """
    compare_roberta_base('roberta-large')
    """
    ### paddlenlp:
    paddle:  The Ġman Ġworked Ġas Ġa [Ġmechanic=0.08259474486112595||Ġdriver=0.057356804609298706||Ġteacher=0.04709123820066452||Ġbartender=0.046422384679317474||Ġwaiter=0.042401134967803955] .
    
    ### huggingface:
    [
    {
        "sequence": "The man worked as a mechanic.",
        "score": 0.08260270208120346,
        "token": 25682,
        "token_str": " mechanic"
    },
    {
        "sequence": "The man worked as a driver.",
        "score": 0.057361237704753876,
        "token": 1393,
        "token_str": " driver"
    },
    {
        "sequence": "The man worked as a teacher.",
        "score": 0.04709038510918617,
        "token": 3254,
        "token_str": " teacher"
    },
    {
        "sequence": "The man worked as a bartender.",
        "score": 0.04641640931367874,
        "token": 33080,
        "token_str": " bartender"
    },
    {
        "sequence": "The man worked as a waiter.",
        "score": 0.04239244386553764,
        "token": 38233,
        "token_str": " waiter"
    }
    ]
    """
    compare_roberta_base('sshleifer/tiny-distilroberta-base')
    """
    ### paddlenlp:
    paddle:  The Ġman Ġworked Ġas Ġa [ĠMaul=2.2346663172356784e-05||ELS=2.234388557553757e-05||Ġirregular=2.2234522475628182e-05||ĠDr=2.2128468117443845e-05||³=2.2114423700259067e-05] .
    
    ### huggingface:
    [
        {
            "sequence": "The man worked as a mechanic.",
            "score": 0.08260270208120346,
            "token": 25682,
            "token_str": " mechanic"
        },
        {
            "sequence": "The man worked as a driver.",
            "score": 0.057361237704753876,
            "token": 1393,
            "token_str": " driver"
        },
        {
            "sequence": "The man worked as a teacher.",
            "score": 0.04709038510918617,
            "token": 3254,
            "token_str": " teacher"
        },
        {
            "sequence": "The man worked as a bartender.",
            "score": 0.04641640931367874,
            "token": 33080,
            "token_str": " bartender"
        },
        {
            "sequence": "The man worked as a waiter.",
            "score": 0.04239244386553764,
            "token": 38233,
            "token_str": " waiter"
        }
    ]
    """
